nixos/modules/security/permissions-wrappers/permissions-wrapper.c

```diff
@@ -0,0 +1,224 @@
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <unistd.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <dirent.h>
+#include <assert.h>
+#include <errno.h>
+
+// Make sure assertions are not compiled out, we use them to codify
+// invariants about this program and we want it to fail fast and
+// loudly if they are violated.
+#undef NDEBUG
+
+extern char **environ;
+
+// The SOURCE_PROG and WRAPPER_DIR macros are supplied at compile time
+// for a security reason: So they cannot be changed at runtime.
+static char * sourceProg = SOURCE_PROG;
+static char * wrapperDir = WRAPPER_DIR;
+
+// Make sure we have the WRAPPER_TYPE macro specified at compile
+// time...
```

Comment:
Do you think that may be a problem with these programs? By unfetchable are you talking about a cache miss or something else? So far, my use of the setcap-wrapper work in production hasn't revealed any unexpected cache misses...

That's a good point. I think the only way to do it would be to check _if_ there are capabilities set on the file. If so, raise them, otherwise do nothing with the capabilities. It's late here so I'll think on this a bit more tomorrow but that does feel cleaner to me.

Thanks for pointing out the run-time error. I've switched that to use the `#error` directive.